Use StringView in CSSParserToken and simplify its code

cdumez opened the pull request:
#### 855b9cf6ef5a2b8b5ead0d06314779d7fd1884fc
<pre>
Use StringView in CSSParserToken and simplify its code
<a href="https://bugs.webkit.org/show_bug.cgi?id=262498">https://bugs.webkit.org/show_bug.cgi?id=262498</a>

Reviewed by Darin Adler.

Use StringView in CSSParserToken and simplify its code.

Previously, CSSParserToken wasn&apos;t using StringView internally for better bit packing,
as the size of this object matters a lot. Instead, do better bit packing in StringView
so that CSSParserToken can use it without increasing its size (24 bits on my
macBookPro M1).

To better bit-pack StringView, we rely on PackedPtr to fit the is8Bit flag in the same
word as the character pointer.

This is performance neutral on Speedometer.

* Source/WTF/wtf/text/StringView.cpp:
(WTF::StringView::convertToASCIILowercase const):
(WTF::StringView::convertToASCIIUppercase const):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::StringView):
(WTF::StringView::characters8 const):
(WTF::StringView::characters16 const):
* Source/WebCore/css/CSSVariableData.cpp:
(WebCore::CSSVariableData::updateBackingStringsInTokens):
* Source/WebCore/css/parser/CSSParserToken.cpp:
(WebCore::CSSParserToken::CSSParserToken):
(WebCore::CSSParserToken::convertToDimensionWithUnit):
(WebCore::CSSParserToken::copyWithUpdatedString const):
* Source/WebCore/css/parser/CSSParserToken.h:
(WebCore::CSSParserToken::value const):
(WebCore::CSSParserToken::updateString):
(WebCore::CSSParserToken::initValueFromStringView): Deleted.
(WebCore::CSSParserToken::updateCharacters): Deleted.

Canonical link: <a href="https://commits.webkit.org/268760@main">https://commits.webkit.org/268760@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03100cfa4c7077014c25c7b5aeb66f2e75ade67b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20595 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21013 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21665 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22485 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19221 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20831 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24267 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21191 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20582 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20816 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20655 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17905 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23342 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17821 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18712 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24985 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/17931 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18900 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18889 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22926 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/20015 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19470 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/16548 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/24003 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18690 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5730 "Found 4 jsc stress test failures: stress/sampling-profiler-microtasks.js.default, stress/sampling-profiler-microtasks.js.eager-jettison-no-cjit, stress/sampling-profiler-microtasks.js.mini-mode, stress/sampling-profiler-microtasks.js.no-llint") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4942 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23026 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/25261 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19289 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5555 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/text/StringView.h:
```diff
@@ -230,13 +229,13 @@ class StringView final {
     UnderlyingString* m_underlyingString { nullptr };
 #endif
+
+    PackedPtr<const void> m_characters;
```
Note that I also tried:
```
CompactPointerType<const void*, bool /* is8Bit */> m_characters;
unsigned m_length;
```

and 
```
const void* m_characters;
unsigned m_length : 31 { 0 };
unsigned m_is8Bit : 1 { true };
```

But only this approach with the `PackedPtr<>` managed to keep the size of `CSSParserToken` at 24.

darinadler replied:
I have no idea what `PackedPtr` does.

darinadler commented on Source/WebCore/css/parser/CSSParserToken.cpp:
```diff
@@ -41,6 +41,29 @@
+        HashTokenType m_hashTokenType;
+        double m_numericValue;
+        mutable int m_id;
+    };
```
I would just put `double m_value` here.

darinadler commented on Source/WebCore/css/CSSVariableData.cpp:
```diff
@@ -48,7 +48,7 @@ template<typename CharacterType> void CSSVariableData::updateBackingStringsInTok
             continue;
         unsigned length = token.value().length();
-        token.updateCharacters(currentOffset, length);
+        token.updateString(StringView { currentOffset, length });
```
I don’t think you need to write `StringView` here. Just the braces should suffice.

darinadler commented on Source/WebCore/css/parser/CSSParserToken.cpp:
```diff
@@ -41,6 +41,29 @@
 namespace WebCore {
 DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(CSSParserToken);
 
+#ifdef NDEBUG
```
Would be nice to get this check even in debug builds. Is there some reason that’s difficult?

cdumez replied:
It failed on the debug bots and it was not immediately clear to me why but I can investigate.

cdumez committed:
Use StringView in CSSParserToken and simplify its code
https://bugs.webkit.org/show_bug.cgi?id=262498

Reviewed by Darin Adler.

Use StringView in CSSParserToken and simplify its code.

Previously, CSSParserToken wasn't using StringView internally for better bit packing,
as the size of this object matters a lot. Instead, do better bit packing in StringView
so that CSSParserToken can use it without increasing its size (24 bits on my
macBookPro M1).

To better bit-pack StringView, we rely on PackedPtr to fit the is8Bit flag in the same
word as the character pointer.

This is performance neutral on Speedometer.

* Source/WTF/wtf/text/StringView.cpp:
(WTF::StringView::convertToASCIILowercase const):
(WTF::StringView::convertToASCIIUppercase const):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::StringView):
(WTF::StringView::characters8 const):
(WTF::StringView::characters16 const):
* Source/WebCore/css/CSSVariableData.cpp:
(WebCore::CSSVariableData::updateBackingStringsInTokens):
* Source/WebCore/css/parser/CSSParserToken.cpp:
(WebCore::CSSParserToken::CSSParserToken):
(WebCore::CSSParserToken::convertToDimensionWithUnit):
(WebCore::CSSParserToken::copyWithUpdatedString const):
* Source/WebCore/css/parser/CSSParserToken.h:
(WebCore::CSSParserToken::value const):
(WebCore::CSSParserToken::updateString):
(WebCore::CSSParserToken::initValueFromStringView): Deleted.
(WebCore::CSSParserToken::updateCharacters): Deleted.

Canonical link: https://commits.webkit.org/268760@main
